File: solutions/python/reverse_linked_list_iterative.py
# Reverse_linked_list_iterative.h 69dab7578339012270872ba200cfff02f59ce894
# Reverse_linked_list_iterative.cpp 848813e190b1b85a8e75107fe8513c3be38ad1a9
from linked_list_prototype import ListNode


# @include
def reverse_linked_list(head):
    prev = None
    curr = head
    while curr:
        nxt = curr.next
        curr.next = prev
        prev = curr
        curr = nxt
    return prev


# @exclude


def print_list(head):
    if head:
        print('(', head.data, ')', sep='')
        print_list(head.next)


def main():
    L1 = ListNode(1, ListNode(2, ListNode(3, None)))

    print('before reverse')
    print_list(L1)
    newhead = reverse_linked_list(L1)
    print('\nafter reverse')
    assert newhead.data == 3 and newhead.next.data == 2 and newhead.next.next.data == 1
    print_list(newhead)
    newhead = reverse_linked_list(newhead)
    assert newhead.data == 1 and newhead.next.data == 2 and newhead.next.next.data == 3
    print('\nafter another reverse')
    print_list(newhead)


if __name__ == '__main__':
    main()

File: solutions/cpp/reverse_linked_list_iterative.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <cassert>
#include <iostream>
#include <memory>

#include "./Linked_list_prototype.h"
#include "./Reverse_linked_list_iterative.h"

using std::cout;
using std::endl;
using std::make_shared;
using std::shared_ptr;

template <typename T>
void Print(shared_ptr<ListNode<T>> head) {
  if (head) {
    cout << "(" << head->data << ")" << endl;
    Print(head->next);
  }
}

int main(int argc, char* argv[]) {
  shared_ptr<ListNode<int>> L1 =
      make_shared<ListNode<int>>(ListNode<int>{1, nullptr});
  shared_ptr<ListNode<int>> L2 =
      make_shared<ListNode<int>>(ListNode<int>{2, nullptr});
  L1->next = L2;
  shared_ptr<ListNode<int>> L3 =
      make_shared<ListNode<int>>(ListNode<int>{3, nullptr});
  L2->next = L3;

  cout << "before reverse" << endl;
  Print(L1);
  shared_ptr<ListNode<int>> newhead = ReverseLinkedList(L1);
  cout << endl << "after reverse" << endl;
  assert(newhead->data == 3 && newhead->next->data == 2 &&
         newhead->next->next->data == 1);
  Print(newhead);
  newhead = ReverseLinkedList(newhead);
  assert(newhead->data == 1 && newhead->next->data == 2 &&
         newhead->next->next->data == 3);
  cout << endl << "after another reverse" << endl;
  Print(newhead);
  return 0;
}
